File: HikvisionService/Controllers/ApiController.cs
using Microsoft.AspNetCore.Mvc;
using HikvisionService.Models;
using HikvisionService.Services;
using HikvisionService.Models.ViewModels;
using HikvisionService.Data;

namespace HikvisionService.Controllers;

[ApiController]
[Route("api")]
public class ApiController : ControllerBase
{
    private readonly IHikvisionService _hikvisionService;
    private readonly ILogger<ApiController> _logger;

    public ApiController(IHikvisionService hikvisionService, ILogger<ApiController> logger)
    {
        _hikvisionService = hikvisionService;
        _logger = logger;
    }

    // Dashboard data
    [HttpGet("dashboard")]
    public async Task<ActionResult<DashboardViewModel>> GetDashboardData()
    {
        try
        {
            var dashboardData = await _hikvisionService.GetDashboardDataAsync();
            return Ok(dashboardData);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting dashboard data");
            return StatusCode(500, new { error = "Failed to get dashboard data" });
        }
    }

    // Camera endpoints
    [HttpGet("cameras")]
    public async Task<ActionResult<List<Camera>>> GetAllCameras()
    {
        try
        {
            var cameras = await _hikvisionService.GetAllCamerasAsync();
            return Ok(cameras);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting all cameras");
            return StatusCode(500, new { error = "Failed to get cameras" });
        }
    }

    [HttpGet("cameras/{id}")]
    public async Task<ActionResult<Camera>> GetCamera(long id)
    {
        try
        {
            var camera = await _hikvisionService.GetCameraByIdAsync(id);
            if (camera == null)
            {
                return NotFound(new { error = $"Camera with ID {id} not found" });
            }
            return Ok(camera);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting camera {CameraId}", id);
            return StatusCode(500, new { error = $"Failed to get camera {id}" });
        }
    }

    [HttpPost("cameras/{id}/check")]
    public async Task<ActionResult> CheckCameraConnection(long id)
    {
        try
        {
            bool isOnline = await _hikvisionService.CheckCameraConnectionAsync(id);
            return Ok(new { isOnline });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error checking camera {CameraId} connection", id);
            return StatusCode(500, new { error = $"Failed to check camera {id} connection" });
        }
    }

    [HttpPost("cameras/check-all")]
    public async Task<ActionResult> CheckAllCameras()
    {
        try
        {
            await _hikvisionService.TriggerCameraHealthCheckAsync();
            return Ok(new { message = "Camera health check triggered" });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error triggering camera health check");
            return StatusCode(500, new { error = "Failed to trigger camera health check" });
        }
    }

    // Storage drive endpoints
    [HttpGet("storage-drives")]
    public async Task<ActionResult<List<StorageDrive>>> GetAllStorageDrives()
    {
        try
        {
            var drives = await _hikvisionService.GetAllStorageDrivesAsync();
            return Ok(drives);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting all storage drives");
            return StatusCode(500, new { error = "Failed to get storage drives" });
        }
    }

    [HttpGet("storage-drives/{id}")]
    public async Task<ActionResult<StorageDrive>> GetStorageDrive(long id)
    {
        try
        {
            var drive = await _hikvisionService.GetStorageDriveByIdAsync(id);
            if (drive == null)
            {
                return NotFound(new { error = $"Storage drive with ID {id} not found" });
            }
            return Ok(drive);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting storage drive {DriveId}", id);
            return StatusCode(500, new { error = $"Failed to get storage drive {id}" });
        }
    }

    [HttpPost("storage-drives")]
    public async Task<ActionResult<StorageDrive>> AddStorageDrive(StorageDrive drive)
    {
        try
        {
            var newDrive = await _hikvisionService.AddStorageDriveAsync(drive);
            return CreatedAtAction(nameof(GetStorageDrive), new { id = newDrive.Id }, newDrive);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error adding storage drive");
            return StatusCode(500, new { error = "Failed to add storage drive" });
        }
    }

    [HttpPut("storage-drives/{id}")]
    public async Task<ActionResult> UpdateStorageDrive(long id, StorageDrive drive)
    {
        if (id != drive.Id)
        {
            return BadRequest(new { error = "ID mismatch" });
        }

        try
        {
            bool success = await _hikvisionService.UpdateStorageDriveAsync(drive);
            if (!success)
            {
                return NotFound(new { error = $"Storage drive with ID {id} not found" });
            }
            return NoContent();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating storage drive {DriveId}", id);
            return StatusCode(500, new { error = $"Failed to update storage drive {id}" });
        }
    }

    [HttpDelete("storage-drives/{id}")]
    public async Task<ActionResult> DeleteStorageDrive(long id)
    {
        try
        {
            bool success = await _hikvisionService.DeleteStorageDriveAsync(id);
            if (!success)
            {
                return NotFound(new { error = $"Storage drive with ID {id} not found" });
            }
            return NoContent();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deleting storage drive {DriveId}", id);
            return StatusCode(500, new { error = $"Failed to delete storage drive {id}" });
        }
    }

    [HttpPost("storage-drives/check")]
    public async Task<ActionResult> CheckStorageDrives()
    {
        try
        {
            await _hikvisionService.TriggerStorageCheckAsync();
            return Ok(new { message = "Storage check triggered" });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error triggering storage check");
            return StatusCode(500, new { error = "Failed to trigger storage check" });
        }
    }

    // Download job endpoints
    [HttpGet("download-jobs")]
    public async Task<ActionResult<List<FileDownloadJobDto>>> GetAllDownloadJobs()
    {
        try
        {
            var jobs = await _hikvisionService.GetAllDownloadJobsAsync();
            return Ok(jobs);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting all download jobs");
            return StatusCode(500, new { error = ex.Message });
        }
    }

    [HttpGet("download-jobs/active")]
    public async Task<ActionResult<List<FileDownloadJobDto>>> GetActiveDownloadJobs()
    {
        try
        {
            var jobs = await _hikvisionService.GetActiveDownloadJobsAsync();
            return Ok(jobs);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting active download jobs");
            return StatusCode(500, new { error = "Failed to get active download jobs" });
        }
    }

    [HttpGet("download-jobs/failed")]
    public async Task<ActionResult<List<FileDownloadJobDto>>> GetFailedDownloadJobs()
    {
        try
        {
            var jobs = await _hikvisionService.GetFailedDownloadJobsAsync();
            return Ok(jobs);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting failed download jobs");
            return StatusCode(500, new { error = "Failed to get failed download jobs" });
        }
    }

    [HttpGet("download-jobs/{id}")]
    public async Task<ActionResult<FileDownloadJob>> GetDownloadJob(long id)
    {
        try
        {
            var job = await _hikvisionService.GetDownloadJobByIdAsync(id);
            if (job == null)
            {
                return NotFound(new { error = $"Download job with ID {id} not found" });
            }
            return Ok(job);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting download job {JobId}", id);
            return StatusCode(500, new { error = $"Failed to get download job {id}" });
        }
    }

    [HttpPost("download-jobs/{id}/retry")]
    public async Task<ActionResult> RetryDownloadJob(long id)
    {
        try
        {
            bool success = await _hikvisionService.RetryDownloadJobAsync(id);
            if (!success)
            {
                return NotFound(new { error = $"Download job with ID {id} not found" });
            }
            return Ok(new { message = $"Download job {id} retry triggered" });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrying download job {JobId}", id);
            return StatusCode(500, new { error = $"Failed to retry download job {id}" });
        }
    }

    [HttpPost("download-jobs/{id}/cancel")]
    public async Task<ActionResult> CancelDownloadJob(long id)
    {
        try
        {
            bool success = await _hikvisionService.CancelDownloadJobAsync(id);
            if (!success)
            {
                return NotFound(new { error = $"Download job with ID {id} not found" });
            }
            return Ok(new { message = $"Download job {id} cancelled" });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error cancelling download job {JobId}", id);
            return StatusCode(500, new { error = $"Failed to cancel download job {id}" });
        }
    }
    
    // Footage endpoints
    [HttpGet("footage")]
    public async Task<ActionResult<List<FootageFileViewModel>>> GetFootageFiles(
        [FromQuery] long? cameraId,
        [FromQuery] DateTime startDate,
        [FromQuery] DateTime endDate,
        [FromQuery] string fileType = "both")
    {
        try
        {
            var files = await _hikvisionService.GetFootageFilesAsync(cameraId, startDate, endDate, fileType);
            return Ok(files);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting footage files");
            return StatusCode(500, new { error = "Failed to get footage files" });
        }
    }
    
    [HttpGet("footage/download")]
    public async Task<ActionResult> DownloadFootage([FromQuery] string path)
    {
        try
        {
            var filePath = await _hikvisionService.GetFootageDownloadUrlAsync(path);
            
            // Get file info
            var fileInfo = new System.IO.FileInfo(filePath);
            if (!fileInfo.Exists)
            {
                return NotFound(new { error = "File not found" });
            }
            
            // Determine content type
            string contentType = "application/octet-stream";
            if (filePath.EndsWith(".mp4", StringComparison.OrdinalIgnoreCase))
            {
                contentType = "video/mp4";
            }
            else if (filePath.EndsWith(".jpg", StringComparison.OrdinalIgnoreCase) ||
                     filePath.EndsWith(".jpeg", StringComparison.OrdinalIgnoreCase))
            {
                contentType = "image/jpeg";
            }
            
            // Return the file
            return PhysicalFile(filePath, contentType, fileInfo.Name);
        }
        catch (FileNotFoundException)
        {
            return NotFound(new { error = "File not found" });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error downloading file: {Path}", path);
            return StatusCode(500, new { error = "Failed to download file" });
        }
    }
}

File: laravel/app/Http/Controllers/Api/ApiController.php
<?php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use App\Models\Camera;
use App\Models\FileDownloadJob;
use App\Models\StorageDrive;
use App\Services\HikvisionService;
use Carbon\Carbon;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Log;
use Illuminate\Support\Facades\Storage;
use Symfony\Component\HttpFoundation\BinaryFileResponse;

class ApiController extends Controller
{
    protected $hikvisionService;

    public function __construct(HikvisionService $hikvisionService)
    {
        $this->hikvisionService = $hikvisionService;
    }

    // Dashboard data
    public function getDashboardData(): JsonResponse
    {
        try {
            $dashboardData = $this->hikvisionService->getDashboardData();
            return response()->json($dashboardData);
        } catch (\Exception $ex) {
            Log::error('Error getting dashboard data', ['error' => $ex->getMessage()]);
            return response()->json(['error' => 'Failed to get dashboard data'], 500);
        }
    }

    public function getWorkerStatus(): JsonResponse
    {
        try {
            $dashboardData = $this->hikvisionService->getDashboardData();
            return response()->json($dashboardData['worker_status']);
        } catch (\Exception $ex) {
            Log::error('Error getting worker status', ['error' => $ex->getMessage()]);
            return response()->json(['error' => 'Failed to get worker status'], 500);
        }
    }

    // Camera endpoints
    public function getAllCameras(): JsonResponse
    {
        try {
            $cameras = Camera::with('store')->get();
            return response()->json($cameras);
        } catch (\Exception $ex) {
            Log::error('Error getting all cameras', ['error' => $ex->getMessage()]);
            return response()->json(['error' => 'Failed to get cameras'], 500);
        }
    }

    public function getCamera($id): JsonResponse
    {
        try {
            $camera = Camera::with('store')->find($id);
            if (!$camera) {
                return response()->json(['error' => "Camera with ID {$id} not found"], 404);
            }
            return response()->json($camera);
        } catch (\Exception $ex) {
            Log::error("Error getting camera {$id}", ['error' => $ex->getMessage()]);
            return response()->json(['error' => "Failed to get camera {$id}"], 500);
        }
    }

    public function checkCameraConnection($id): JsonResponse
    {
        try {
            $isOnline = false;
            $results = $this->hikvisionService->testCameraConnection($id);
            foreach ($results as $result) {
                if (strpos($result, 'successful') !== false) {
                    $isOnline = true;
                }
            }
            return response()->json(['isOnline' => $isOnline]);
        } catch (\Exception $ex) {
            Log::error("Error checking camera {$id} connection", ['error' => $ex->getMessage()]);
            return response()->json(['error' => "Failed to check camera {$id} connection"], 500);
        }
    }

    public function checkAllCameras(): JsonResponse
    {
        try {
            // Simulate health check for all cameras
            $cameras = Camera::all();
            foreach ($cameras as $camera) {
                $this->hikvisionService->testCameraConnection($camera->id);
            }

            return response()->json(['message' => 'Camera health check triggered']);
        } catch (\Exception $ex) {
            Log::error('Error triggering camera health check', ['error' => $ex->getMessage()]);
            return response()->json(['error' => 'Failed to trigger camera health check'], 500);
        }
    }

    // Storage drive endpoints
    public function getAllStorageDrives(): JsonResponse
    {
        try {
            $drives = StorageDrive::all();
            return response()->json($drives);
        } catch (\Exception $ex) {
            Log::error('Error getting all storage drives', ['error' => $ex->getMessage()]);
            return response()->json(['error' => 'Failed to get storage drives'], 500);
        }
    }

    public function getStorageDrive($id): JsonResponse
    {
        try {
            $drive = StorageDrive::find($id);
            if (!$drive) {
                return response()->json(['error' => "Storage drive with ID {$id} not found"], 404);
            }
            return response()->json($drive);
        } catch (\Exception $ex) {
            Log::error("Error getting storage drive {$id}", ['error' => $ex->getMessage()]);
            return response()->json(['error' => "Failed to get storage drive {$id}"], 500);
        }
    }

    public function addStorageDrive(Request $request): JsonResponse
    {
        try {
            $request->validate([
                'name' => 'required|string|max:255',
                'root_path' => 'required|string|max:500',
            ]);

            $drive = StorageDrive::create([
                'name' => $request->name,
                'root_path' => $request->root_path,
                'total_space' => 0,
                'used_space' => 0,
                'free_space' => 0,
                'status' => 'unknown',
                'last_checked_at' => now(),
            ]);

            return response()->json($drive, 201);
        } catch (\Exception $ex) {
            Log::error('Error adding storage drive', ['error' => $ex->getMessage()]);
            return response()->json(['error' => 'Failed to add storage drive'], 500);
        }
    }

    public function updateStorageDrive(Request $request, $id): JsonResponse
    {
        try {
            $drive = StorageDrive::find($id);
            if (!$drive) {
                return response()->json(['error' => "Storage drive with ID {$id} not found"], 404);
            }

            $request->validate([
                'name' => 'required|string|max:255',
                'root_path' => 'required|string|max:500',
            ]);

            $drive->update([
                'name' => $request->name,
                'root_path' => $request->root_path,
            ]);

            return response()->json(null, 204);
        } catch (\Exception $ex) {
            Log::error("Error updating storage drive {$id}", ['error' => $ex->getMessage()]);
            return response()->json(['error' => "Failed to update storage drive {$id}"], 500);
        }
    }

    public function deleteStorageDrive($id): JsonResponse
    {
        try {
            $drive = StorageDrive::find($id);
            if (!$drive) {
                return response()->json(['error' => "Storage drive with ID {$id} not found"], 404);
            }

            $drive->delete();
            return response()->json(null, 204);
        } catch (\Exception $ex) {
            Log::error("Error deleting storage drive {$id}", ['error' => $ex->getMessage()]);
            return response()->json(['error' => "Failed to delete storage drive {$id}"], 500);
        }
    }

    public function checkStorageDrives(): JsonResponse
    {
        try {
            $this->hikvisionService->checkStorageDrives();
            return response()->json(['message' => 'Storage check triggered']);
        } catch (\Exception $ex) {
            Log::error('Error triggering storage check', ['error' => $ex->getMessage()]);
            return response()->json(['error' => 'Failed to trigger storage check'], 500);
        }
    }

    // Download job endpoints
    public function getAllDownloadJobs(): JsonResponse
    {
        try {
            $jobs = FileDownloadJob::with('camera')
                ->orderByDesc('created_at')
                ->get()
                ->map(function ($job) {
                    return [
                        'id' => $job->id,
                        'status' => $job->status,
                        'progress' => $job->progress,
                        'created_at' => $job->created_at,
                        'start_time' => $job->start_time,
                        'end_time' => $job->end_time,
                        'updated_at' => $job->updated_at,
                        'camera_id' => $job->camera_id,
                        'camera_name' => $job->camera->name,
                        'file_name' => $job->file_name,
                        'file_type' => $job->file_type,
                        'file_size' => $job->file_size,
                        'error_message' => $job->error_message,
                    ];
                });

            return response()->json($jobs);
        } catch (\Exception $ex) {
            Log::error('Error getting all download jobs', ['error' => $ex->getMessage()]);
            return response()->json(['error' => $ex->getMessage()], 500);
        }
    }

    public function getActiveDownloadJobs(): JsonResponse
    {
        try {
            $jobs = FileDownloadJob::with('camera')
                ->where('status', 'downloading')
                ->orderByDesc('start_time')
                ->get()
                ->map(function ($job) {
                    return [
                        'id' => $job->id,
                        'status' => $job->status,
                        'progress' => $job->progress,
                        'created_at' => $job->created_at,
                        'start_time' => $job->start_time,
                        'end_time' => $job->end_time,
                        'updated_at' => $job->updated_at,
                        'camera_id' => $job->camera_id,
                        'camera_name' => $job->camera->name,
                        'file_name' => $job->file_name,
                        'file_type' => $job->file_type,
                        'file_size' => $job->file_size,
                        'error_message' => $job->error_message,
                    ];
                });

            return response()->json($jobs);
        } catch (\Exception $ex) {
            Log::error('Error getting active download jobs', ['error' => $ex->getMessage()]);
            return response()->json(['error' => 'Failed to get active download jobs'], 500);
        }
    }

    public function getFailedDownloadJobs(): JsonResponse
    {
        try {
            $jobs = FileDownloadJob::with('camera')
                ->where('status', 'failed')
                ->orderByDesc('updated_at')
                ->get()
                ->map(function ($job) {
                    return [
                        'id' => $job->id,
                        'status' => $job->status,
                        'progress' => $job->progress,
                        'created_at' => $job->created_at,
                        'start_time' => $job->start_time,
                        'end_time' => $job->end_time,
                        'updated_at' => $job->updated_at,
                        'camera_id' => $job->camera_id,
                        'camera_name' => $job->camera->name,
                        'file_name' => $job->file_name,
                        'file_type' => $job->file_type,
                        'file_size' => $job->file_size,
                        'error_message' => $job->error_message,
                    ];
                });

            return response()->json($jobs);
        } catch (\Exception $ex) {
            Log::error('Error getting failed download jobs', ['error' => $ex->getMessage()]);
            return response()->json(['error' => 'Failed to get failed download jobs'], 500);
        }
    }

    public function getDownloadJob($id): JsonResponse
    {
        try {
            $job = FileDownloadJob::with('camera')->find($id);
            if (!$job) {
                return response()->json(['error' => "Download job with ID {$id} not found"], 404);
            }
            return response()->json($job);
        } catch (\Exception $ex) {
            Log::error("Error getting download job {$id}", ['error' => $ex->getMessage()]);
            return response()->json(['error' => "Failed to get download job {$id}"], 500);
        }
    }

    // Footage endpoints
    public function getFootageFiles(Request $request): JsonResponse
    {
        try {
            $cameraId = $request->query('cameraId');
            $startDate = Carbon::parse($request->query('startDate', now()->subDays(7)));
            $endDate = Carbon::parse($request->query('endDate', now()));
            $fileType = $request->query('fileType', 'both');

            $files = $this->hikvisionService->getFootageFiles($cameraId, $startDate, $endDate, $fileType);
            return response()->json($files);
        } catch (\Exception $ex) {
            Log::error('Error getting footage files', ['error' => $ex->getMessage()]);
            return response()->json(['error' => 'Failed to get footage files'], 500);
        }
    }

    public function downloadFootage(Request $request): BinaryFileResponse|JsonResponse
    {
        try {
            $path = $request->query('path');
            if (!$path) {
                return response()->json(['error' => 'Path parameter is required'], 400);
            }

            // For demo purposes, return a placeholder response
            // In real implementation, this would serve the actual file
            if (!file_exists($path)) {
                return response()->json(['error' => 'File not found'], 404);
            }

            // Determine content type
            $contentType = 'application/octet-stream';
            if (str_ends_with(strtolower($path), '.mp4')) {
                $contentType = 'video/mp4';
            } elseif (str_ends_with(strtolower($path), '.jpg') || str_ends_with(strtolower($path), '.jpeg')) {
                $contentType = 'image/jpeg';
            }

            return response()->file($path, [
                'Content-Type' => $contentType,
                'Content-Disposition' => 'attachment; filename="' . basename($path) . '"',
            ]);
        } catch (\Exception $ex) {
            Log::error("Error downloading file: {$request->query('path')}", ['error' => $ex->getMessage()]);
            return response()->json(['error' => 'Failed to download file'], 500);
        }
    }
}
